### PR TITLE
add ConstantinopleFix

### PR DESCRIPTION
**Pull request by winsvega**
tests in progress

### CODE REVIEW

File: libethashseal/GenesisInfo.h [hunk 1]
```diff
@@ -65,13 +65,15 @@ enum class Network
     ByzantiumNoProofTest = 81,
     /// Constantinople rules + NoProof seal engine
     ConstantinopleNoProofTest = 82,
+    /// Byzantium + Constantinople + Constantinople FIx acrive from block 0
```
Comment (gumb0) [inline libethashseal/GenesisInfo.h]:
```suggestion
    /// Byzantium + Constantinople + ConstantinopleFix active from block 0
```

File: libethashseal/genesis/test/ByzantiumToConstantinopleAt5Test.cpp [hunk 2]
```diff
@@ -16,8 +16,8 @@ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 #include "../../GenesisInfo.h"
 
-static std::string const c_genesisInfoByzantiumToConstantinopleAt5Test = std::string() +
-                                                                         R"E(
+static std::string const c_genesisInfoByzantiumToConstantinopleFixAt5Test = std::string() +
```
Comment (gumb0) [inline libethashseal/genesis/test/ByzantiumToConstantinopleAt5Test.cpp]: I guess we should better rename this file, too

File: libethashseal/genesis/test/constantinopleTestFix.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU General Public License for more details.
+
+        You should have received a copy of the GNU General Public License
+        along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "../../GenesisInfo.h"
+
+static std::string const c_genesisInfoConstantinopleTestFix = std::string() +
```
Comment (gumb0) [inline libethashseal/genesis/test/constantinopleTestFix.cpp]: It should be `c_genesisInfoConstantinopleFixTest`. Fix file name, too

File: test/tools/libtesteth/TestHelper.cpp [hunk 4]
```diff
@@ -120,7 +122,7 @@ string netIdToString(eth::Network _netId)
         return "HomesteadToEIP150At5";
     case eth::Network::EIP158ToByzantiumAt5:
         return "EIP158ToByzantiumAt5";
-    case eth::Network::ByzantiumToConstantinopleAt5:
+    case eth::Network::ByzantiumToConstantinopleFixAt5:
         return "ByzantiumToConstantinopleAt5";
```
Comment (gumb0) [inline test/tools/libtesteth/TestHelper.cpp]: Why not change this?

**Comment by winsvega** [comment]
hm linux-gcc6 not passing the tests while macos is finishing all the tests no issues.

File: aleth-vm/main.cpp [hunk 5]
```diff
@@ -226,6 +226,8 @@ int main(int argc, char** argv)
     if (vm.count("network"))
     {
         string network = vm["network"].as<string>();
+        if (network == "ConstantinopleFix")
```
Comment (gumb0) [inline aleth-vm/main.cpp]: If you're adding it to this tool, then add this possible value to help output, on line 131

File: libethashseal/GenesisInfo.h [hunk 6]
```diff
@@ -65,13 +65,15 @@ enum class Network
     ByzantiumNoProofTest = 81,
     /// Constantinople rules + NoProof seal engine
     ConstantinopleNoProofTest = 82,
+    /// Byzantium + Constantinople + ConstantinopleFix acrive from block 0
```
Comment (gumb0) [inline libethashseal/GenesisInfo.h]:
```suggestion
    /// Byzantium + Constantinople + ConstantinopleFix active from block 0
```

File: libethashseal/genesis/test/constantinopleFixTest.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+        You should have received a copy of the GNU General Public License
+        along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "../../GenesisInfo.h"
+
+static std::string const c_genesisInfoConstantinopleFixTest = std::string() +
+                                                              R"E(
+{
+	"sealEngine": "Ethash",
```
Comment (gumb0) [inline libethashseal/genesis/test/constantinopleFixTest.cpp]: convert tabs to spaces

File: test/tools/jsontests/BlockChainTests.cpp [hunk 8]
```diff
@@ -763,10 +763,10 @@ void overwriteUncleHeaderForTest(mObject& uncleHeaderObj, TestBlock& uncle, std:
         return;
     }
 
-    string overwrite = "false";
```
Comment (gumb0) [inline test/tools/jsontests/BlockChainTests.cpp]: Is it necessary to have these testeth logic modifications in the PR introducing ConstantinopleFix?
It should better go to separate PR

Comment (winsvega) [reply test/tools/jsontests/BlockChainTests.cpp]: Some blockchain test filling was failing again. This as a part of the solution to refill the blockchain tests.

File: libethashseal/genesis/test/ByzantiumToConstantinopleFixAt5Test.cpp [hunk 9]
```diff
@@ -16,8 +16,8 @@ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 */
 #include "../../GenesisInfo.h"
 
-static std::string const c_genesisInfoByzantiumToConstantinopleAt5Test = std::string() +
-                                                                         R"E(
+static std::string const c_genesisInfoByzantiumToConstantinopleFixAt5Test = std::string() +
```
Comment (gumb0) [inline libethashseal/genesis/test/ByzantiumToConstantinopleFixAt5Test.cpp]: But this config doesn't have `constantinopleFixForkBlock`, why?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 10]
```diff
@@ -809,42 +809,45 @@ void overwriteUncleHeaderForTest(mObject& uncleHeaderObj, TestBlock& uncle, std:
         uncleHeader = uncle.blockHeader();
     }
 
-    if (overwrite != "false")
+    if (overwrite.size())
```
Comment (gumb0) [inline test/tools/jsontests/BlockChainTests.cpp]: this check is redundant, as the loop will do nothing if it's empty anyway

**Comment by gumb0** [comment]
Linux clang build on CI runs tests with aleth-interpreter, which you didn't update for `ConstantinopleFix` here, that's the reason for failures.

@chfast I guess we should add aleth-interpreter changes into this PR, too.

**Comment by gumb0** [comment]
Rebased and added a fix for aleth-interpreter.

File: libaleth-interpreter/VM.cpp [hunk 11]
```diff
@@ -1360,12 +1360,13 @@ void VM::interpretCases()
                 m_runGas = VMSchedule::sstoreSetGas;
             else if (status == EVMC_STORAGE_MODIFIED || status == EVMC_STORAGE_DELETED)
                 m_runGas = VMSchedule::sstoreResetGas;
-            else if (status == EVMC_STORAGE_UNCHANGED && m_rev < EVMC_CONSTANTINOPLE)
+            else if (status == EVMC_STORAGE_UNCHANGED &&
+                     (m_rev < EVMC_CONSTANTINOPLE || m_rev >= EVMC_CONSTANTINOPLE2))
```
Comment (chfast) [inline libaleth-interpreter/VM.cpp]:
```suggestion
                     m_rev != EVMC_CONSTANTINOPLE
```

File: libaleth-interpreter/VM.cpp [hunk 12]
```diff
@@ -1360,12 +1360,13 @@ void VM::interpretCases()
...
                 m_runGas = VMSchedule::sstoreResetGas;
             else
             {
                 assert(status == EVMC_STORAGE_UNCHANGED || status == EVMC_STORAGE_MODIFIED_AGAIN);
-                assert(m_rev >= EVMC_CONSTANTINOPLE);
+                assert(m_rev >= EVMC_CONSTANTINOPLE && m_rev < EVMC_CONSTANTINOPLE2);
```
Comment (chfast) [inline libaleth-interpreter/VM.cpp]:
```suggestion
                assert(m_rev == EVMC_CONSTANTINOPLE);
```

File: test/unittests/libtesteth/testHelperTest.cpp [hunk 13]
```diff
@@ -28,6 +28,14 @@ using namespace dev::test;
 
 BOOST_FIXTURE_TEST_SUITE(TestHelperSuite, TestOutputHelperFixture)
 
+BOOST_AUTO_TEST_CASE(translateNetworks_gtConstantinople)
+{
+    set<string> networks = {">Constantinople"};
+    networks = test::translateNetworks(networks);
+    BOOST_REQUIRE(networks.count("Constantinople") == 0);
```
Comment (chfast) [inline test/unittests/libtesteth/testHelperTest.cpp]: Use `BOOST_CHECK_EQ`.

**Comment by gumb0** [comment]
@chfast issues addressed
